jxl/src/frame/modular/transforms/apply.rs

```diff
@@ -134,3 +141,484 @@ impl TransformStepChunk {
+    Ok(())
+}
+
+fn meta_apply_single_transform(
+    transform: &headers::modular::Transform,
+    channels: &mut Vec<(usize, ChannelInfo)>,
+    transform_steps: &mut Vec<TransformStep>,
+    mut add_transform_buffer: impl FnMut(ChannelInfo, String) -> usize,
+) -> Result<()> {
+    match transform.id {
+        TransformId::Rct => {
+            let begin_channel = transform.begin_channel as usize;
+            let op = RctOp::from_u32(transform.rct_type % 7).unwrap();
+            let perm = RctPermutation::from_u32(transform.rct_type / 7)
+                .expect("header decoding should ensure rct_type < 42");
+            check_equal_channels(channels, begin_channel, 3)?;
+            let mut buf_in = [0; 3];
+            let buf_out = [
+                channels[begin_channel].0,
+                channels[begin_channel + 1].0,
+                channels[begin_channel + 2].0,
+            ];
+            for i in 0..3 {
+                let c = &mut channels[begin_channel + i];
+                c.0 = add_transform_buffer(
+                    c.1,
+                    format!(
+                        "RCT (op {:?} perm {:?}) starting at channel {}, input {}",
+                        op, perm, begin_channel, i
+                    ),
+                );
+                buf_in[i] = c.0;
+            }
+            transform_steps.push(TransformStep::Rct {
+                buf_out,
+                buf_in,
+                op,
+                perm,
+            });
+            trace!("applied RCT: {channels:?}");
+        }
+        TransformId::Squeeze => {
+            let steps = if transform.squeezes.is_empty() {
+                super::squeeze::default_squeeze(channels)
+            } else {
+                transform.squeezes.clone()
+            };
+            for step in steps {
+                super::squeeze::check_squeeze_params(channels, &step)?;
+                let in_place = step.in_place;
+                let horizontal = step.horizontal;
+                let begin_channel = step.begin_channel as usize;
+                let num_channels = step.num_channels as usize;
+                let end_channel = begin_channel + num_channels;
+                let new_chan_offset = if in_place {
+                    end_channel
+                } else {
+                    channels.len()
+                };
+                for ic in 0..num_channels {
+                    let chan = &channels[begin_channel + ic].1;
+                    let new_shift = if let Some(shift) = chan.shift {
+                        if shift.0 > 30 || shift.1 > 30 {
+                            return Err(Error::TooManySqueezes);
+                        }
+                        if horizontal {
+                            Some((shift.0 + 1, shift.1))
+                        } else {
+                            Some((shift.0, shift.1 + 1))
+                        }
+                    } else {
+                        None
+                    };
```

Comment:
Isn’t this:

```suggestion
                    let new_shift = chan.shift.map(|shift| {
                        if shift.0 > 30 || shift.1 > 30 {
                            return Err(Error::TooManySqueezes);
                        }
                        if horizontal {
                            Some((shift.0 + 1, shift.1))
                        } else {
                            Some((shift.0, shift.1 + 1))
                        }
                    });
```

Reply:
No, the `return` inside `map` doesn't quite work right :-) this would need a `transpose()` or some other mildly painful additional step